File: device-app/android/Java/umsDemo/app/src/main/java/com/rationalowl/umsdemo/domain/MessagesRepository.java
package com.rationalowl.umsdemo.domain;

import android.util.Log;

import com.jakewharton.rxrelay3.BehaviorRelay;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;

public class MessagesRepository extends JsonRepository<Message[]> {
    private static final String TAG = "MessagesRepository";
    private static final String FILE_NAME = "messages.json";

    private static MessagesRepository instance;

    public static synchronized MessagesRepository getInstance() {
        if (instance == null) {
            instance = new MessagesRepository();
        }

        return instance;
    }

    private final BehaviorRelay<List<Message>> messages;
    private final Map<String, Message> messageIdMappings = new HashMap<>();

    protected MessagesRepository() {
        super(Message[].class, FILE_NAME);

        final Message[] array = getValue();
        final List<Message> list = new ArrayList<>();

        if (array != null) {
            final Date now = new Date();
            final int retainDay = Config.getInstance().getUmsMsgRetainDay();

            for (Message message : array) {
                long milliseconds = now.getTime() - message.getSentAt().getTime();
                long days = TimeUnit.DAYS.convert(milliseconds, TimeUnit.MILLISECONDS);

                if (days < retainDay) {
                    list.add(message);
                    messageIdMappings.put(message.getId(), message);
                }
            }

            list.sort((m1, m2) -> m2.getSentAt().compareTo(m1.getSentAt()));
        }

        messages = BehaviorRelay.createDefault(list);

        if (array != null && array.length != list.size()) {
            save();
        }
    }

    public void addMessage(Message item) {
        Log.d(TAG, "addMessage(" + item.getId() + ")");

        if (hasMessage(item.getId())) return;

        synchronized (this) {
            final List<Message> newItems = new ArrayList<>(messages.getValue());
            newItems.add(item);
            newItems.sort((m1, m2) -> m2.getSentAt().compareTo(m1.getSentAt()));
            messageIdMappings.put(item.getId(), item);

            messages.accept(newItems);
        }

        save();
    }

    public synchronized boolean hasMessage(String messageId) {
        return messageIdMappings.containsKey(messageId);
    }

    public synchronized Message getMessage(String messageId) {
        return messageIdMappings.get(messageId);
    }

    public BehaviorRelay<List<Message>> getMessages() {
        return messages;
    }

    public void removeMessage(String messageId) {
        Log.d(TAG, "removeMessage(" + messageId + ")");

        synchronized (this) {
            final List<Message> newItems = new ArrayList<>(messages.getValue());
            newItems.removeIf(message -> message.getId().equals(messageId));
            messageIdMappings.remove(messageId);

            messages.accept(newItems);
        }

        save();
    }

    public void removeMessages(String[] messageIds) {
        Log.d(TAG, "removeMessage([" + String.join(", ", messageIds) + "])");

        synchronized (this) {
            final List<Message> newItems = new ArrayList<>(messages.getValue());
            final Set<String> messageIdSet = new HashSet<>(Arrays.asList(messageIds));

            newItems.removeIf(message -> messageIdSet.contains(message.getId()));

            for (String messageId : messageIds) {
                messageIdMappings.remove(messageId);
            }

            messages.accept(newItems);
        }

        save();
    }

    public void setAsRead(String messageId) {
        Log.d(TAG, "setAsRead(" + messageId + ")");

        synchronized (this) {
            final List<Message> newItems = new ArrayList<>(messages.getValue());

            try {
                final Message oldItem = messageIdMappings.get(messageId);
                if (oldItem == null) return;

                final Message newItem = oldItem.clone();
                newItem.setAsRead();
                newItems.set(newItems.indexOf(oldItem), newItem);

                messageIdMappings.put(messageId, newItem);
            } catch (CloneNotSupportedException e) {
                e.printStackTrace();
            }

            messages.accept(newItems);
        }

        save();
    }

    public void setAllAsRead() {
        Log.d(TAG, "setAllAsRead()");

        synchronized (this) {
            final List<Message> newItems = new ArrayList<>();

            for (Message item : messages.getValue()) {
                try {
                    final Message newItem = item.clone();
                    newItem.setAsRead();
                    newItems.add(newItem);

                    messageIdMappings.put(item.getId(), newItem);
                } catch (CloneNotSupportedException e) {
                    e.printStackTrace();
                }
            }

            messages.accept(newItems);
        }

        save();
    }

    public synchronized void save() {
        Log.d(TAG, "save()");
        setValue(messages.getValue().toArray(new Message[0]));
    }
}

File: device-app/ios-swift/umsDemo/ums/Domain/MessagesRepository.swift
import RxCocoa
import RxSwift

class MessagesRepository : JsonRepository<[Message]>
{
    private static let fileName = "messages.json"
    
    static let shared = MessagesRepository()
    
    let messages: BehaviorRelay<[Message]> = BehaviorRelay(value: [])
    private var messageIdMappings = [String: Message]()
    
    private let lock = NSLock()
    
    private init()
    {
        super.init(MessagesRepository.fileName)
        
        if let array: [Message] = getValue()
        {
            var list = [Message]()
            
            let now = Date()
            let retainDay: Int = Config.shared?.umsMsgRetainDay ?? 0
            
            for message in array
            {
                let days: Int = Calendar.current.dateComponents([.day], from: message.sentAt, to: now).day!
                
                if days < retainDay
                {
                    list.append(message)
                    messageIdMappings[message.id] = message
                }
            }
            
            messages.accept(list)
            
            if array.count != list.count
            {
                save()
            }
        }
    }
    
    func addMessage(_ item: Message)
    {
        NSLog("[\(type(of: self))] addMessage('\(item.title)')")
        
        if hasMessage(item.id) { return }
        
        lock.lock()
        var newItems = messages.value
        newItems.append(item)
        newItems.sort { $0.sentAt > $1.sentAt }
        messageIdMappings[item.id] = item
        
        messages.accept(newItems)
        lock.unlock()
        
        save()
    }
    
    func hasMessage(_ messageId: String) -> Bool
    {
        lock.lock()
        let result: Bool = messageIdMappings.keys.contains(messageId)
        lock.unlock()
        
        return result
    }
    
    func getMessage(_ messageId: String) -> Message?
    {
        lock.lock()
        let result: Message? = messageIdMappings[messageId]
        lock.unlock()
        
        return result
    }
    
    func removeMessage(_ messageId: String)
    {
        NSLog("[\(type(of: self))] removeMessage(\(messageId))")
        
        lock.lock()
        var newItems = messages.value
        
        if let index = newItems.firstIndex(where: { $0.id == messageId })
        {
            newItems.remove(at: index)
        }
        
        messageIdMappings[messageId] = nil
        
        messages.accept(newItems)
        lock.unlock()
        
        save()
    }
    
    func removeMessages(_ messageIds: [String])
    {
        let messageIdsString = messageIds.joined(separator: ", " )
        NSLog("[\(type(of: self))] removeMessage([\(messageIdsString)])")
        
        lock.lock()
        var newItems = messages.value
        let messageIdSet = Set(messageIds)
        
        newItems.removeAll { messageIdSet.contains($0.id) }
        
        for messageId: String in messageIds
        {
            messageIdMappings[messageId] = nil
        }
        
        messages.accept(newItems)
        lock.unlock()
        
        save()
    }
    
    func setAsRead(_ messageId: String)
    {
        NSLog("[\(type(of: self))] setAsRead(\(messageId))")
        
        lock.lock()
        var newItems = messages.value
        
        if let oldItem = messageIdMappings[messageId]
        {
            var newItem = oldItem
            newItem.isRead = true
            
            if let index = newItems.firstIndex(of: oldItem)
            {
                newItems[index] = newItem
            }
            
            messageIdMappings[messageId] = newItem
        }
        
        messages.accept(newItems)
        lock.unlock()
        
        save()
    }
    
    func setAllAsRead()
    {
        NSLog("[\(type(of: self))] setAllAsRead()")
        
        lock.lock()
        var newItems = [Message]()
        
        for item in messages.value
        {
            var newItem = item
            newItem.isRead = true
            newItems.append(newItem)
            
            messageIdMappings[item.id] = newItem
        }
        
        messages.accept(newItems)
        lock.unlock()
        
        save()
    }
    
    func save()
    {
        NSLog("[\(type(of: self))] save()")
        setValue(messages.value)
    }
}
